---
title: "mbImpute: an accurate and robust imputation method for microbiome data"
author: "Ruochen Jiang, Wei Vivian Li, and Jingyi Jessica Li"

date: "`r Sys.Date()`"
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# mbImpute

<!-- badges: start -->
<!-- badges: end -->

The goal of mbImpute is to impute false zero counts in microbiome sequencing data, i.e., a sample-by-taxon count matrix, by jointly borrowing information from similar samples, similar taxa and optional metadata including sample covariates, and taxon phylogeny.

## Installation

After you download the package zip file, you may install the mbImpute package (e.g., Version 0.1.0) using the following command in R:

``` r
install.packages("mbImpute_0.1.0.tar.gz")
```

You may also use the following R code to directly install the mbImpute package from GitHub: 

``` r
#install.packages("devtools")
library(devtools)
install_github("ruochenj/mbImpute/mbImpute R package")
```

## Example

We use the microbiome dataset from Karlsson et al (2013) as an example to demonstrate the use of mbImpute:

```{r example}
library(mbImpute)
library(glmnet)
# the OTU table
otu_tab[1:6, 1:6]
# the taxon phylogenetic distance matrix 
D[1:6, 1:6]
# the (optional) meta data, i.e., sample covariate matrix with rows representing samples and corresponding to the rows in otu_tab
meta_data[1:6, 1:6]
# obtain the sample conditions from the meta data (imputation will be performed within each condition)
condition = meta_data$study_condition

# For all the categorical variables, make sure they are converted to numerical.
meta_data[,1] <- as.numeric(as.factor(meta_data[,1]))

# meta_data <- meta_data[,-1]

# run mbImpute
imputed_count_mat_list <- mbImpute(condition = condition, otu_tab = otu_tab, meta_data = meta_data, D = D)
# If you have multiple cores and would like to do parallel computing, please use the following command
# imputed_matrix <- mbImpute(condition = condition, otu_tab = otu_tab, meta_data = meta_data, D = D, k =5, parallel = TRUE, ncores = 4)
# If you do not have meta data, or phylogenetic information, and the samples belong to one condition
# otu_tab_T2D <- otu_tab[condition == "T2D",]
# imputed_count_matrix <- mbImpute(otu_tab = otu_tab_T2D)
# a glance at the imputed matrix
# First result is the count matrix at log scale, we recommend to perform downstream analysis on this data as the distribution for the values in each taxon follows approximately normal distribution (see our paper for more results)
imputed_count_mat_list$imputed_count_mat_lognorm[1:3, 1:2]
# Second result is the imputed normalized count matrix with same library size set to 10^6 for each sample (subject/person).
imputed_count_mat_list$imp_count_mat_norm[1:3, 1:2]
# Third result is the imputed count matrix with the same scale as the input count matrix.
imputed_count_mat_list$imp_count_mat_origlibsize[1:3, 1:2]
```

Reference:

Karlsson, F. H., Tremaroli, V., Nookaew, I., Bergström, G., Behre, C. J., Fagerberg, B., ... & Bäckhed, F. (2013). Gut metagenome in European women with normal, impaired and diabetic glucose control. Nature, 498(7452), 99-103.
